### ISSUE TITLE
Research-critical gaps: persistent latent reasoning, LM block semantics, calibrated verification, long-context state, and adaptive compute

### ISSUE THREAD

**Issue by s-b-repo** [s-b-repo/unifur#5]
# Research-critical gaps and flaws remaining in Unifur

This issue records the remaining gaps found by auditing the current implementation against the roadmap and the repository's own claims. These are not merely missing benchmarks; several affect whether the proposed latent-reasoning / multi-axis LM architecture can be meaningfully evaluated.

## 1. Persistent latent reasoning is not implemented

The current LM path is still a conventional causal transformer. `LanguageModel::forward_span()` executes a contiguous layer range, but it does not expose a persistent latent state that can undergo multiple internal refinement steps before emitting the next token.

Implement a first-class latent state API:

```text
prompt -> H0
H0 -> refine -> H1
H1 -> refine -> H2
...
Hk -> verifier / token head -> output
```

Requirements:

- latent state must persist across internal refinement steps;
- internal compute steps must be decoupled from emitted-token count;
- zero refinement should reduce exactly to the baseline;
- refinement depth must be budgeted;
- intermediate states must be inspectable and benchmarkable.

## 2. `forward_span()` does not establish correct end-to-end LM block semantics

The current span execution starts from token embedding for the supplied tokens. That is useful infrastructure, but it does not by itself prove that independently trained spans learn the same input distribution they receive when composed end to end.

Add explicit tests/experiments for:

- full sequential forward vs span-by-span forward;
- training only one span and then composing all spans;
- distribution shift at block boundaries;
- whether block-local loss predicts end-to-end quality;
- whether later blocks should consume detached/teacher representations, live preceding representations, or a latent-state target.

Document the chosen semantics rather than treating image-style block training as automatically transferable to language.

## 3. Diffusion-specific conditioning is inherited into the LM without evidence

The LM reuses the image `DbLayer` and feeds timestep zero through `TimestepEmbedder` even though a normal LM has no diffusion timestep.

This is architecturally convenient but needs an ablation against:

- standard Transformer normalization/conditioning;
- current shared diffusion-style conditioning;
- latent-step conditioning;
- recurrent-state conditioning.

The goal is to establish whether the shared trunk is actually beneficial for language or merely inherited implementation baggage.

## 4. Fixed learned absolute positions are a major long-context limitation

The LM uses a learned position table with a fixed context size. Cached decoding stops when the table is exhausted, while ordinary generation can left-truncate.

For long-context research, add a position/state abstraction supporting:

- longer contexts without reallocating a fixed position table;
- cache-compatible continuation;
- windowed/sliding context;
- recurrent state;
- retrieval from older context.

Compare learned absolute positions with a long-context-compatible alternative before evaluating the multi-axis architecture.

## 5. The cache model is insufficient for adaptive latent computation

The KV cache currently captures attention keys/values, but the planned architecture needs additional persistent state:

```text
attention cache
recurrent memory
latent reasoning state
routing state
verifier state
memory/retrieval state
```

Design a unified inference-state object with explicit ownership, reset, truncation and serialization semantics.

Add equivalence tests when any subset of the state is cached or recomputed.

## 6. Adaptive depth uses confidence without a correctness-calibrated verifier

Confidence/halting is not equivalent to correctness. Before enabling adaptive compute as a quality optimization, implement a correctness-calibrated verifier or halting head.

Measure:

- false early exits;
- false continuations;
- calibration error;
- quality loss at fixed compute savings;
- quality/latency Pareto curves.

Do not use raw confidence as a proxy for correctness without evidence.

## 7. Consistency can collapse latent blocks into identity mappings

Consistency encourages adjacent states to agree, but useful reasoning requires measurable progress.

Add a latent progress objective, for example a margin over provisional prediction quality:

```text
score(H_{t+1}) >= score(H_t) + margin
```

or an equivalent verifier-derived progress target.

Test whether later latent states improve correctness rather than merely increasing similarity to earlier states.

## 8. Parallel latent branches need anti-collapse and complementary-computation objectives

The proposed architecture allows:

```text
H0 -> B1/B2/B3/B4 -> aggregator -> H1
```

but there is currently no mechanism proving that branches perform complementary computation rather than four copies of the same transformation.

Implement and compare:

- mean aggregation;
- learned weighted sum;
- attention aggregation;
- confidence-weighted aggregation;
- gated mixture.

Add branch diversity / usefulness diagnostics and an objective that avoids trivial branch collapse without forcing arbitrary diversity.

## 9. Planner is search infrastructure, not yet a learned compute-policy controller

The Phase 21 planner searches candidate paths under a budget. The multi-axis architecture requires a policy that estimates expected value of compute actions.

Add a deterministic compute-policy interface first, then a learned value function:

```text
policy = P(h, routing_state, budget)
value(action) = expected_quality_gain / expected_cost
```

Candidate actions should include:

- attention mode;
- memory read;
- MoVA top-k;
- FFN top-k;
- extra latent depth;
- verification;
- MTP width.

All actions must expose measurable compute cost.

## 10. MoE routing still lacks the strongest routing-health mechanisms

The roadmap still identifies incomplete work around:

- global-batch load balancing;
- loss-free bias balancing;
- routing entropy diagnostics.

Finish these before building a multi-axis router, otherwise routing behavior can be confounded by micro-batch composition and expert collapse.

Track separately:

- marginal load entropy;
- per-token routing entropy;
- routing stability across adjacent layers/tokens;
- expert starvation;
- dispatch overhead.

## 11. No rigorous evidence yet that parallelism improves quality/compute

The repository has execution and benchmark infrastructure, but several decisive quality comparisons remain GPU-blocked.

A real validation matrix must compare at equal:

- training steps;
- examples/tokens seen;
- FLOPs;
- wall-clock time;
- GPU-hours;
- parameter count / active parameters.

The primary metric should be quality per unit compute, not raw optimizer-step speed.

## 12. No rigorous evidence yet that adaptive depth improves the Pareto frontier

The implementation can skip/widen computation, but the remaining experiment is whether this produces a real dominance over fixed-depth baselines.

Required curves:

```text
quality vs average compute
quality vs p50 latency
quality vs p95 latency
quality vs expected depth
```

Keep adaptive computation only if it gives a measured Pareto improvement.

## 13. No rigorous evidence that consistency, MoE, flow matching, or distillation improve final quality

The mechanisms are implemented and many invariants are tested, but implementation is not evidence of final-task superiority.

Each mechanism needs isolated ablations with repeated seeds and raw measurements.

## 14. End-to-end verification is still weaker than component certificates

The repository has a strong numerical certificate discipline, but historical bugs show that a certificate can accidentally validate an independently reimplemented formula rather than the real execution path.

For every important claim, add a layered verification chain:

1. mathematical identity;
2. implementation-level property;
3. end-to-end behavioral invariant;
4. controlled benchmark.

Prefer tests that call the production implementation directly.

## 15. Checkpoint state must cover the new adaptive/latent architecture

The existing roadmap emphasizes model/optimizer/scheduler/RNG/data state. The latent architecture additionally needs reproducible state for:

- latent refinement controller;
- halting/verifier state;
- routing state;
- memory state;
- planner state;
- cached attention state when applicable.

Add resume-equivalence tests that include adaptive computation decisions, not just final weights.

## 16. Research-gating rule

Do not scale directly to the proposed 24–70B multi-axis systems.

Required progression:

### Stage A
Dense LM baseline.

### Stage B
Persistent latent state + sequential refinement.

### Stage C
Progress objective + calibrated verifier/halting.

### Stage D
Parallel latent branches.

### Stage E
Adaptive compute controller.

### Stage F
MoVA / latent-MoE / hybrid attention.

### Stage G
Long-context recurrent/retrieval memory.

### Stage H
MTP and full multi-axis system.

Only scale after a smaller model demonstrates a reproducible quality/compute improvement.

## 17. Central falsification experiment

The most important experiment for the new direction is:

> At matched training/inference compute and matched visible output length, does allowing additional internal latent refinement steps improve reasoning/task accuracy compared with simply adding the equivalent compute to ordinary token-emitting transformer depth?

Required variants:

1. baseline fixed-depth autoregressive LM;
2. deeper fixed-depth LM with matched compute;
3. sequential latent refinement;
4. parallel latent refinement;
5. latent refinement + progress objective;
6. latent refinement + verifier/halting.

Measure:

- task accuracy;
- validation loss/perplexity;
- visible tokens generated;
- latent steps used;
- FLOPs;
- p50/p95 latency;
- memory;
- robustness across seeds.

## Definition of done

This issue is complete only when the project can distinguish clearly between:

- mechanisms that merely exist;
- mechanisms that satisfy local mathematical invariants;
- mechanisms that improve end-to-end behavior;
- mechanisms that improve quality per unit compute.

The core target is not architectural complexity. It is a demonstrated improvement in useful computation **between token emissions**, with evidence that the computation is productive rather than redundant or merely self-consistent.
